docs/bos/queryapi/context.md

```diff
@@ -0,0 +1,265 @@
+
+#### Input
+
+```js
+await context.db.upsert(objects, conflictColumns, updateColumns)
+```
+
+The Objects parameter is either one or an array of objects. The other two parameters are arrays of strings. The strings *should* correspond to column names for that table. 
+
+#### Example
+
+```js
+const insertPostDataA = {
+  id: postId
+  account_id: accountIdA,
+  block_height: blockHeightA,
+  block_timestamp: blockTimeStampA
+};
+
+
+const insertPostDataB = {
+  id: postId
+  account_id: accountIdB,
+  block_height: blockHeightB,
+  block_timestamp: blockTimeStampB
+};
+// Insert new post to Posts table
+await context.db.Posts.upsert(
+  [ insertPostDataA, insertPostDataB ], 
+  [‘account_id’, ‘id’], 
+  [‘block_height’, ‘block_timestamp’);
+```
+
+In this example, two objects are being inserted. However, if a row already exists in the table where the account_id and id are the same, then block_height and block_timestamp would be overwritten in those rows to the value in the object in the call which is conflicting. 
+
+### Delete
+
+This method deletes all objects in the row which exactly match the object passed in. Obviously caution should be taken when using this method. I currently only support AND and exact match, just like in the select method here. That doubles as a safety measure against accidentally deleting a bunch of data. Later on, I may add more functionality related to matching, as well as limits. All deleted rows ARE returned so you can always insert them back if you get back more rows than expected or something. Or reindex your indexer I guess. 
+
+#### Input
+
+```js
+await context.db.TableName.delete(whereObj)
+```
+
+As stated, only a single object is allowed. 
+
+#### Example
+
+```js
+await context.db.delete_post_likes(
+  { 
+    account_id: likeAuthorAccountId, 
+    post_id: postId
+  }
+);
+```
+
+In this example, any rows where account_id and post_id match the supplied value are deleted. All deleted rows are returned. 
+
+### Auto Complete
+
+The above methods are generated when the schema is read. The tables in the schema are parsed and methods set under each table name. This makes using the object more intuitive and declarative. In addition to that, typescript types are generated which represent the table itself. These types are set as the parameter’s types. This provides autocomplete and strong typing in the IDE. These restrictions are not enforced on the runner side and are instead mainly as a suggestion to help guide the developer to use the methods in a way that is deemed correct by us. Here’s some screenshots of autocomplete and strong typing in action: 
```

Comment:
Pavel gave some good advice that we should highlight that the autocomplete works while writing the schema and before publishing to the chain. In other words, you don't need to publish the indexer to get autocomplete. 

Some additions of my own: 
Right now, indexers by default had a context object included as a parameter in the top-level "async function getBlock". This blocks autocomplete as the local context object will shadow the global one, preventing access to it. Customers will need to manually remove the context paramter from their indexers and they will then see autocomplete. 

`async function getBlock(block: Block, context) {`

should become

`async function getBlock(block: Block) {`

I also added a button next to the format button. It has a "<>" symbol in it. That manually generates the types. It can be useful for debugging and iterative development while writing the schema.

Reply:
Also forgot to add, if the schema is invalid for generating types, then an error will pop up both on screen and in the console. Here's a picture of that. 
<img width="830" alt="image" src="https://github.com/near/docs/assets/22734869/1540d7d3-9f81-4034-bebf-1c80ac65d45f">

<img width="614" alt="image" src="https://github.com/near/docs/assets/22734869/c393eba7-77ab-4c83-ad1a-d7f9660911a4">

Reply:
Hi @darunrs , thanks for the additional notes and screenshots

I have updated the PR , adding all these notes. Please review it when you have a chance 👍